File: Kinematics.py
import numpy as np
import copy
import scipy.optimize 
import sys
from Constants import *





# Get the rotation matrix for two given vectors.
def rotationMatrix(v1, v2):
    # Get the cross product of the two vectors.
    k = np.cross(v1, v2)
    # Get the identity matrix.
    I = np.identity(3)
   
    # Test if the norm of the cross product is zero.
    if np.linalg.norm(k) == 0:
        return I
    else:
        # Unit vector of k.
        k = k / np.linalg.norm(k)
        # Get the K matrix.
        K = np.array([[0, -k[2], k[1]], 
                      [k[2], 0, -k[0]],
                      [-k[1], k[0], 0]])
        # Get the angle between the two vectors.
        theta = np.arccos(np.clip(np.dot(v1/np.linalg.norm(v1), v2/ np.linalg.norm(v2)), -1, 1))
        
        # Get the rotation matrix.
        RotMat = I + np.sin(theta) * K + (1 - np.cos(theta)) *np.dot(K, K)
        return RotMat

# Function to check the momentum conservation.
def check_mom_cons(Event):
    total = [0, 0, 0]
    # Go through the list of events
    for pm in Event:
        total[0] += pm.Px
        total[1] += pm.Py
        total[2] += pm.Pz
    
    return total

# Rotate a given particle with a given rotation matrix.
def rotate(p, rotMat):
    
    pvec = [p.Px, p.Py, p.Pz]
    rotvec = np.dot(rotMat, pvec)
    
    rotp = copy.deepcopy(p)
    rotp.Px = rotvec[0]
    rotp.Py = rotvec[1]
    rotp.Pz = rotvec[2]
    return  rotp

# Rotate the given particles to allign with the mother particle's momentum in the lab frame
def RotateMomentaLab(p, particles):
    
    rotated_particles = []
    
    pmag = np.sqrt(p.Px**2 + p.Py**2 + p.Pz**2)
    Matrix = rotationMatrix([0, 0, pmag], [p.Px, p.Py, p.Pz])
    for p in particles:
        
        i = copy.deepcopy(p)
        v = np.array([i.Px, i.Py, i.Pz])
        rotedVec = np.dot(Matrix, v)
        i.Px = rotedVec[0]
        i.Py = rotedVec[1]
        i.Pz = rotedVec[2]
        rotated_particles.append(i)
    return rotated_particles

# Get the boost factor beta for the outgouting jet (new) and parent jet (old)
def Boost_factor(k, new, old):

    qs = new[0]**2 + new[1]**2 + new[2]**2
    q = np.sqrt(qs)
    Q2 = new[3]**2 - new[0]**2 - new[1]**2 - new[2]**2
    kp = k* np.sqrt(old[0]**2 + old[1]**2 + old[2]**2)
    kps= kp**2
    betamag = (q * new[3] - kp * np.sqrt(kps + Q2)) / (kps + qs + Q2)
    beta = betamag * (k / kp) * np.array([old[0], old[1], old[2]])

    if betamag >= 0:
        return beta
    else:
        return [0, 0, 0]    

# The function to boost a given particle with a given boost factor beta
def boost(p, beta):
    
    bmag = np.sqrt(beta[0]**2 + beta[1]**2 + beta[2]**2)
    # Get the gamma factor
    gamma = 1 / np.sqrt(1 - bmag**2)

    boosted_particle = copy.deepcopy(p)
    
    # Use the matrix of a lorentz boost from https://www.physicsforums.com/threads/general-matrix-representation-of-lorentz-boost.695941/
    boosted_particle.Px = - gamma * beta[0] * p.E + (    1 + (gamma -1) * beta[0]**2 / bmag**2) * p.Px + ( (gamma - 1) * beta[0] * beta[1] / bmag**2) * p.Py +  ((gamma - 1) * beta[0] * beta[2] / bmag**2) * p.Pz
    boosted_particle.Py = - gamma * beta[1] * p.E + ( (gamma -1) * beta[0] * beta[1] / bmag**2) * p.Px + (    1 + (gamma - 1) * beta[1]**2 / bmag**2) * p.Py + ( (gamma - 1) * beta[1] * beta[2] / bmag**2) * p.Pz
    boosted_particle.Pz = - gamma * beta[2] * p.E + ( (gamma -1) * beta[0] * beta[2] / bmag**2) * p.Px + ( (gamma - 1) * beta[2] * beta[1] / bmag**2) * p.Py + (     1 + (gamma - 1) * beta[2]**2/ bmag**2) * p.Pz
    boosted_particle.E =    gamma * p.E - gamma * beta[0] * p.Px - gamma * beta[1] * p.Py - gamma * beta[2] * p.Pz

    return boosted_particle

# The equation to numerically solve to find k
def K_eq(k, p, q, s):
    sump = 0
    # Loop through the momentum and sum up each term
    for i in range(len(p)):
        sump = sump + np.sqrt(k*k *p[i] + q[i])
        
    sump = sump - s
    
    return sump

# Solve for the k factor
def Solve_k_factor(pj, qj, s):
    
    kargs = (pj, qj, s)
    k = scipy.optimize.root(K_eq, 1.01, args = kargs)
    
    return k.x[0]

# Define the function to perfom global momentum conservation on the jets and particles
def Glob_mom_cons(ShoweredParticles, Jets):
    
    pj = [] # The momenta of the parent parton
    qj = [] # The momenta of the jets
    newqs = [] # Array to hold the outgoing jet's momentum
    oldps = [] # Array to hold the progenitor's momentum
    rotms = [] # The rotatio matrices
    # Initialize the total energy
    sqrts = 0
    

    # Iterate through the list of jets
    for jet in Jets:
        
        # Append the 3-momentum of the Jet's progenitor to the 
        pj.append(jet.Progenitor.Px**2 + jet.Progenitor.Py**2 + jet.Progenitor.Pz**2)
        
        # Get the jet progenitor's 4 momentum
        oldp = np.array([jet.Progenitor.Px, jet.Progenitor.Py, jet.Progenitor.Pz, jet.Progenitor.E])
          
        # Add this jet's progenitor's energy
        sqrts += jet.Progenitor.E
        
        # Get the total jet's momentum after showering
        newq = [0, 0, 0, 0]
        # Iterate through the particles in the jet and add their momentum
        for p in jet.Particles:
            newp = copy.deepcopy(p)
            newq[0] = newq[0] + newp.Px
            newq[1] = newq[1] + newp.Py
            newq[2] = newq[2] + newp.Pz
            newq[3] = newq[3] + newp.E
        
        # Calculate the momentum and append it to the list of jet momentum
        qj2 = newq[3]**2 - newq[0]**2 - newq[1]**2 - newq[2]**2
        if np.isnan(qj2):
            qj2 = 0
        Rqp = rotationMatrix(np.array([newq[0], newq[1], newq[2]]), np.array([oldp[0], oldp[1], oldp[2]] ))
        
        oldps.append(oldp)
        newqs.append(newq)
        qj.append(qj2)
        rotms.append(Rqp)
        
    # Get the k factor
    k = Solve_k_factor(pj, qj, sqrts)
    # List to hold the showered particles
    showered_particles = []
    # Go through the list of particles and add back the inital state
    for p in ShoweredParticles:
        if abs(p.typ) == 11:
            showered_particles.append(p)

    # Check if any of the jets have been radiated and if so, do not perform the boost and append the progenitor of each jet
    if_ratiated = any(len(jet.Particles) > 1 for jet in Jets)
    if if_ratiated is False:
        for i, j in enumerate(Jets):
            showered_particles.append(j.Particles[0])
     
    # Iterate though the jets list and boost each particle
    else:
        for i, jet in enumerate(Jets):

            # Get boosted factor
            beta = Boost_factor(k, newqs[i], oldps[i])
            # Get a copy of the jet
            # This is for redudancy so the original jet's information is not modified 
            jet = copy.deepcopy(jet)
            # Go through the jet's particles and rotate and boost each
            for p in jet.Particles:
                rotated = rotate(p, rotms[i])

                # Get the boosted particle
                pboosted = boost(rotated, beta)
                # Append to showered particles
                showered_particles.append(pboosted)
                
    return showered_particles

File: juliaPartonShower/Kinematics.jl
include("Constants.jl")

using Roots
using LinearAlgebra

# Get the rotation matrix for two given vectors
function getRotationMatrix(v1::Vector{Float64}, v2::Vector{Float64})
    # The the cross product of two vectors
    k = cross(v1, v2)

    if norm(k) == 0
        return [1 0 0; 0 1 0; 0 0 1]
    else
        # Get unit vector of k
        k = k / norm(k)
        # Get the k matrix
        K = [0 -k[3] k[2]; k[3] 0 -k[1]; -k[2] k[1] 0]
        
        theta = acos(clamp(dot(v1/norm(v1), v2/norm(v2)), -1, 1))

        # Get the rotation matrix
        rotMat = I + sin(theta) * K + (1 - cos(theta)) * K^2
        return rotMat
    end
end

# Function to rotate a given particle with a given rotation matrix
function rotate(p::Particle, rotMat::Matrix)
    pvec = [p.px, p.py, p.pz]
    rotVec = rotMat * pvec
    p.px = rotVec[1]
    p.py = rotVec[2]
    p.pz = rotVec[3]
    return p
end


# Function to rotate the particle to allign with the mother particle's momentum in the lab frame
function rotateMomentaLab(p::Particle, particles::Vector{Particle})

    rotatedParticles = []

    pmag = sqrt(p.px^2 + p.py^2 + p.pz^2)
    m = getRotationMatrix([0, 0, pmag], [p.px, p.py, p.pz])
    for p in particles
        v = [p.px, p.py, p.pz]
        rotatedVec = m * v
        p.px = rotatedVec[1]
        p.py = rotatedVec[2]
        p.pz = rotatedVec[3]
        push!(rotatedParticles, p)

    end
    return rotatedParticles
end

# Function to get the boost factor for the outgoing jet (new) and parent jet (old)
function boostFactor(k::Float64, new::Vector{Float64}, old::Vector{Float64})
    
    qs = new[1]^2 + new[2]^2 + new[3]^2
    q = sqrt(qs)
    Q2 = new[4]^2 - new[1]^2 - new[2]^2 - new[3]^2
    kp = k * sqrt(old[1]^2 + old[2]^2 + old[3]^2)
    kps = kp^2
    betamag = (q * new[4] - kp * sqrt(kps + Q2)) / (kps + qs + Q2)
    beta = betamag * (k / kp) * [old[1], old[2], old[3]]

    if betamag >= 0
        return beta
    else
        return [0, 0, 0]
    end

end


function boost(p::Particle, beta::Vector{Float64})

    bmag = sqrt(beta[1]^2 + beta[2]^2 + beta[3]^2)
    gamma = 1/ sqrt(1 - bmag^2)

    p.px = - gamma * beta[0] * p.E + (    1 + (gamma -1) * beta[0]^2 / bmag^2) * p.px + ( (gamma - 1) * beta[0] * beta[1] / bmag^2) * p.py +  ((gamma - 1) * beta[0] * beta[2] / bmag^2) * p.pz
    p.py = - gamma * beta[1] * p.E + ( (gamma -1) * beta[0] * beta[1] / bmag^2) * p.px + (    1 + (gamma - 1) * beta[1]^2 / bmag^2) * p.py + ( (gamma - 1) * beta[1] * beta[2] / bmag^2) * p.pz
    p.pz = - gamma * beta[2] * p.E + ( (gamma -1) * beta[0] * beta[2] / bmag^2) * p.px + ( (gamma - 1) * beta[2] * beta[1] / bmag^2) * p.py + (     1 + (gamma - 1) * beta[2]^2/ bmag^2) * p.pz
    p.E =    gamma * p.E - gamma * beta[0] * p.px - gamma * beta[1] * p.py - gamma * beta[2] * p.pz

    return p
end

# The function to numerically solve to find k
function kEq(k::Float64, p::Vector{Float64}, q::Vector{Float64}, s::Float64)
    sump = 0
    for i in range(1, length(p))

        sump = sump + sqrt(k^2 * p[i] + q[i])
        
    end
    sump = sump - s
    return sump
end

# Function to numerically solve for k
function solvekFactor(pj::Vector{Float64}, qj::Vector{Float64}, s::Float64)
    # Turn the function into one that can be parsed with the arguements into the root finder
    kEQ = (k -> kEq(k, pj, qj, s))
    sol = find_zero(kEQ, 1.01)
    return sol
end


# Function to perform global momemntum conservation
function globalMomCons(showeredParticles::Vector{Particle}, Jets::Vector{Jet})
    pj = Float64[] # The momenta of the parent parton
    qj = Float64[] # The momenta of the jets
    newqs = Vector{Float64}[] # Array to hold the outgoing jet's momentum
    oldps = Vector{Float64}[] # The progenitor's momentum
    rotms = Matrix{Float64}[] # The rotation matrices
    # Initialize the total energy
    sqrts = 0


    for jet in Jets
        # Append the 3-momentum of the Jet's progentior
        append!(pj, jet.Progenitor.px^2 + jet.Progenitor.py^2 + jet.Progenitor.pz^2)

        # Get the jet progenitor's 4 momentum
        oldp = [jet.Progenitor.px, jet.Progenitor.py, jet.Progenitor.pz, jet.Progenitor.E]

        # Add this jet's progenitor's energy
        sqrts += jet.Progenitor.E

        # Get the total jet's momentum after showering
        newq = [0, 0, 0, 0]
        for p in jet.AllParticles
            newq += [p.px, p.py, p.pz, p.E]
        end

        # Calculate the momentum and append it to the list of jet momentum
        qj2 = newq[4]^2 - newq[1]^2 - newq[2]^2 - newq[3]^2
        if isnan(qj2)
            qj2 = 0
        end
        rotMat = getRotationMatrix([newq[1], newq[2], newq[3]], [oldp[1], oldp[2], oldp[3]])
        push!(oldps, oldp)
        push!(newqs, newq)
        append!(qj, qj2) 
        push!(rotms, rotMat)
    end

    # Get the k factor
    k = solvekFactor(pj, qj, sqrts)

    rotatedShoweredParticles = []

    # Get the electron's and postirons and append them to the list of roated showered particles since these don't need to be rotated
    for p in showeredParticles
        if abs(p.id) == 11
            push!(rotatedShoweredParticles, p)
        end
    end

    # Check if any of the jets have been radiated, if not, do not perform the boost and append the progenitor of each jet
    ifRadiated = any(length(jet.AllParticles) > 1 for jet in Jets)
    if ifRadiated == false
        for j in Jets
            push!(rotatedShoweredParticles, j.AllParticles[1])
        end
    else
        for (i, jet) in enumerate(Jets)
            
            # Get the boost factor
            beta = boostFactor(k, newqs[i], oldps[i])

            # Iterate through the jet's particles and rotate and boost each one
            for p in jet.AllParticles
                rotated = rotated(p, rotms[i])
                pboosted = boost(rotated, beta)
                push!(rotatedShoweredParticles, pboosted)

            end
        end
    end
    return rotatedShoweredParticles
end


# Function to reconstruct the sudakov basis for the entire tree with the progenitor being the starting root 
function reconSudakovBasis(prog::Particle, progPart::Particle)
    if prog.status == 1
        return
    end
    current = prog.children[1]
    parent = prog
    stack = [prog]

    # If the stack is empty and the currently selected particle has no children, i.e. final state, then stop the tree search
    while length(stack) != 0 || current.status == -1

        # If the current particle is initial state, i.e. there is an emisison, then append the current particle to the stack
        # then calculate the emisison of the current particle and then select the next particle as the first child in the 
        # list of children.
        if current.status == -1
            push!(stack, current)
            calculatePhysicals(current, prog, progPart, parent)
            current = current.children[1]
        
        # If the current particle is final state, i.e. no emission, then calculate its phyiscals, then select the next particle
        # as the second child of its parent and remove this parent of the stack of particles
        elseif current.status == 1
            calculatePhysicals(current, prog, progPart, parent)
            current = pop!(stack)
            current = current.children[2]
        end

    end


end

# Function to actually calculate the phyiscals of the sudakov basis for a given particle
function calculatePhysicals(part::Particle, prog::Particle, progPart::Particle, parent::Particle)

    pdotn = dot4Vec([prog.px, prog.py, prog.pz, prog.E], [progPart.px, progPart.py, progPart.pz, progPart.pz])

    alpha = parent.alpha * part.z
    part.alpha = alpha
    qi2 = part.virtuality
    kT = [part.pT * cos(part.phi), part.pT * sin(part.phi), 0 , 0] # pT = (px, py, pz, E)
    part.qT = (parent.px * part.z - kT[1], parent.py * part.z - kT[2], 0, 0) # qT = (px, py, pz, E)
    betai = (qi2 - part.alpha^2 * part.m^2 - (-part.qT[1]^2 - part.qT[2])) / (2 * part.alpha * pdotn)

    part.px = alpha * prog.px + betai * progPart.px + part.qT[1]
    part.py = alpha * prog.py + betai * progPart.py + part.qT[2]
    part.pz = alpha * prog.pz + betai * progPart.pz
    part.E = alpha * prog.E + betai * progPart.E

end

function dot4Vec(v1::Vector{Float64}, v2::Vector{Float64})
    return v1[4] * v2[4] - v1[1] * v2[1] - v1[2] * v2[2] - v1[3] * v2[3]
end

# Function to find the color partner of a particle
function findColorPartner(parti::Particle, particles::Vector{Particle})

    partner = parti

    for pc in particles
        if patri.color == pc.antiColor && parti.antiColor == pc.color
            partner = pc
        end
    end 

end
